keydata.go

```diff
@@ -482,110 +505,172 @@ func (d *KeyData) snapModelAuthKey(auxKey AuxiliaryKey) ([]byte, error) {
 
-	stream := cipher.NewCFBEncrypter(c, key[passphraseEncryptionKeyLen:])
-	stream.XORKeyStream(d.data.PassphraseProtectedPayload.EncryptedPayload, payload)
+	stream := cipher.NewCFBEncrypter(c, iv)
+	stream.XORKeyStream(d.data.EncryptedPayload, payload)
 
 	return nil
 }
 
-func (d *KeyData) openWithPassphrase(passphrase string, kdf KDF) (payload []byte, key []byte, err error) {
-	if d.AuthMode()&AuthModePassphrase == 0 {
-		return nil, nil, errors.New("passphrase is not enabled")
+func (d *KeyData) openWithPassphrase(passphrase string, kdf KDF) (payload []byte, authKey []byte, err error) {
+	key, iv, authKey, err := d.derivePassphraseKeys(passphrase, kdf)
+	if err != nil {
+		return nil, nil, err
 	}
 
-	data := d.data.PassphraseProtectedPayload
-	if data.KDF.Type != kdfType {
-		// Only Argon2i is supported
-		return nil, nil, fmt.Errorf("unexpected KDF type \"%s\"", data.KDF.Type)
-	}
-	if data.Encryption != passphraseEncryption {
+	if d.data.PassphraseParams.Encryption != passphraseEncryption {
 		// Only AES-CFB is supported
-		return nil, nil, fmt.Errorf("unexpected encryption algorithm \"%s\"", data.Encryption)
-	}
-	if data.KeySize > 32 {
-		// The key size can't be larger than 32 with the supported cipher
-		return nil, nil, fmt.Errorf("invalid key size (%d bytes)", data.KeySize)
+		return nil, nil, fmt.Errorf("unexpected encryption algorithm \"%s\"", d.data.PassphraseParams.Encryption)
 	}
 
-	// Derive both the key and IV from the passphrase in a single pass.
-	keyLen := data.KeySize + aes.BlockSize
+	payload = make([]byte, len(d.data.EncryptedPayload))
 
-	params := &KDFCostParams{
-		Time:      uint32(data.KDF.Time),
-		MemoryKiB: uint32(data.KDF.Memory),
-		Threads:   uint8(data.KDF.CPUs)}
-	key, err = kdf.Derive(passphrase, data.KDF.Salt, params, uint32(keyLen))
+	c, err := aes.NewCipher(key)
 	if err != nil {
-		return nil, nil, xerrors.Errorf("cannot derive key from passphrase: %w", err)
-	}
-	if len(key) != keyLen {
-		return nil, nil, errors.New("KDF returned unexpected key length")
+		return nil, nil, xerrors.Errorf("cannot create cipher: %w", err)
 	}
+	stream := cipher.NewCFBDecrypter(c, iv)
+	stream.XORKeyStream(payload, d.data.EncryptedPayload)
 
-	payload = make([]byte, len(data.EncryptedPayload))
+	return payload, authKey, nil
+}
 
-	c, err := aes.NewCipher(key[:data.KeySize])
-	if err != nil {
-		return nil, nil, xerrors.Errorf("cannot create cipher: %w", err)
+func (d *KeyData) platformKeyData() *PlatformKeyData {
+	return &PlatformKeyData{
+		Version:       d.Version(),
+		EncodedHandle: d.data.PlatformHandle,
+		KDFAlg:        crypto.Hash(d.data.KDFAlg),
+		AuthMode:      d.AuthMode(),
 	}
-	stream := cipher.NewCFBDecrypter(c, key[data.KeySize:])
-	stream.XORKeyStream(payload, data.EncryptedPayload)
+}
 
-	return payload, key, nil
+func (d *KeyData) recoverKeysCommon(data []byte) (DiskUnlockKey, PrimaryKey, error) {
+	switch d.Version() {
+	case 0:
+		unlockKey, primaryKey, err := unmarshalV1KeyPayload(data)
```

Comment:
This is related to https://github.com/snapcore/secboot/pull/265#discussion_r1450633800 as well.

Reply:
So I already changed it and replied to the comment above but summarizing it again to make sure I don't miss something. The following versions are possible:
- version 0. Key data already used are missing a version field. `Version()` called in that keydata will return 1 and `unmarshalV1KeyPayload` will be called.
- version 1 shouldn't exist (?), but even then the behavior would be the same as version 0 keys. 
- version 2 is the currently introduced format.

When I first saw V1, I assumed there was another V0. Isn't that the case? If not, maybe we could simply call old keys legacy keys?

Reply:
Maybe "generation"? So that we can refer to cross-platform keydata versions as generations and then each platform can refer to its internal version simply as "version". 

i.e TPM platform's key versions 0-2 use cross-platform keys of generation 1 (or 0, this is still a bit confusing) while TPM platform's version 3 keys use generation 2 cross-platform keys.

Reply:
"generation" would work for me

Reply:
Yes, generation is ok with me as well.